Signal/src/view controllers/SignalAttachment.swift

```diff
@@ -0,0 +1,477 @@
+//
+// [SignalAttachment hasError] will be true for non-valid attachments.
+//
+// TODO: Perhaps do conversion off the main thread?
+// TODO: Show error on error.
+// TODO: Show progress on upload.
+class SignalAttachment: NSObject {
+
+    static let TAG = "[SignalAttachment]"
+
+    // MARK: Properties
+
+    let data: Data!
+
+    // Attachment types are identified using UTIs.
+    //
+    // See: https://developer.apple.com/library/content/documentation/Miscellaneous/Reference/UTIRef/Articles/System-DeclaredUniformTypeIdentifiers.html
+    let dataUTI: String!
+
+    var error: SignalAttachmentError? {
+        didSet {
+            AssertIsOnMainThread()
+
+            assert(oldValue == nil)
+            Logger.verbose("\(SignalAttachment.TAG) Attachment has error: \(error)")
+        }
+    }
+
+    // To avoid redundant work of repeatedly compressing/uncompressing
+    // images, we cache the UIImage associated with this attachment if
+    // possible.
+    public var image: UIImage?
+
+    // MARK: Constants
+
+    /**
+     * Media Size constraints from Signal-Android
+     * (org/thoughtcrime/securesms/mms/PushMediaConstraints.java)
+     */
+    static let kMaxFileSizeGif = 5 * 1024 * 1024
+    static let kMaxFileSizeImage = 420 * 1024
+    static let kMaxFileSizeVideo = 100 * 1024 * 1024
+    static let kMaxFileSizeAudio = 100 * 1024 * 1024
+    // TODO: What should the max file size on "other" attachments be?
+    static let kMaxFileSizeGeneric = 100 * 1024 * 1024
+
+    // MARK: Constructor
+
+    // This method should not be called directly; use the factory
+    // methods instead.
+    internal required init(data: Data!, dataUTI: String!) {
+        self.data = data
+        self.dataUTI = dataUTI
+        super.init()
+    }
+
+    public func hasError() -> Bool {
+        return error != nil
+    }
+
+    // Returns the MIME type for this attachment or nil if no MIME type
+    // can be identified.
+    public func mimeType() -> String? {
+        let mimeType = UTTypeCopyPreferredTagWithClass(dataUTI as CFString, kUTTagClassMIMEType)
+        guard mimeType != nil else {
+            return nil
+        }
+        return mimeType?.takeRetainedValue() as? String
+    }
+
+    // Returns the file extension for this attachment or nil if no file extension
+    // can be identified.
+    public func fileExtension() -> String? {
+        let fileExtension = UTTypeCopyPreferredTagWithClass(dataUTI as CFString, kUTTagClassFilenameExtension)
+        guard fileExtension != nil else {
+            return nil
+        }
+        return fileExtension?.takeRetainedValue() as? String
+    }
+
+    // Returns the set of UTIs that correspond to valid _input_ image formats
+    // for Signal attachments.
+    //
+    // Image attachments may be converted to another image format before 
+    // being uploaded.
+    //
+    // TODO: We need to finalize which formats we support.
```

Comment:
I've taken a run at this and `SignalAttachment` now uses `MIMETypeUtil`.

https://github.com/WhisperSystems/SignalServiceKit/pull/125

Still, there's a lot to think over here.  It's worth considering what a "supported" file type means, once we allow users to attach arbitrary files.

For example, why would we impose our current low limits on images (6 MB) but allow users to send 100 MB zip files (which might contain just a single image file)?  Especially in an age where the pictures from phones' cameras are often larger than 6MB?  I'm hoping we can use a single max attachment file size for all attachments - or only apply an exception that is _more generous_, not less, for certain attachment types.

If the app is only concerned with how to _display_ attachments, shouldn't that depend on the capabilities of the current phone, rather than being hard-coded?  Also, neither a UTI nor a MIME type is fine-grained enough to determine whether or not the current phone can play a given .mp4, for example.  It depends on the compression used, the frame rate, the content size, etc.  Maybe that doesn't really matter. http://stackoverflow.com/questions/2719214/what-are-the-video-formats-supported-by-mpmovieplayercontroller-on-the-ipad